File: java/sorting/SelectionSortProduction.java
import java.util.Comparator;
import java.util.List;

public final class SelectionSortProduction {
    private SelectionSortProduction() {
    }

    public static <T extends Comparable<? super T>> void selectionSortInPlace(List<T> values) {
        selectionSortInPlace(values, Comparator.naturalOrder());
    }

    public static <T> void selectionSortInPlace(List<T> values, Comparator<? super T> comparator) {
        if (values == null || comparator == null) {
            throw new IllegalArgumentException("values and comparator must not be null");
        }

        int left = 0;
        int right = values.size() - 1;

        while (left < right) {
            int minIndex = left;
            int maxIndex = left;

            for (int index = left + 1; index <= right; index++) {
                if (comparator.compare(values.get(index), values.get(minIndex)) < 0) {
                    minIndex = index;
                }

                if (comparator.compare(values.get(maxIndex), values.get(index)) < 0) {
                    maxIndex = index;
                }
            }

            swap(values, left, minIndex);

            if (maxIndex == left) {
                maxIndex = minIndex;
            }

            swap(values, right, maxIndex);
            left++;
            right--;
        }
    }

    private static <T> void swap(List<T> values, int left, int right) {
        if (left != right) {
            T temp = values.get(left);
            values.set(left, values.get(right));
            values.set(right, temp);
        }
    }
}

File: python/sorting/selection_sort_production.py
def selection_sort_in_place(values, *, key=None, reverse=False):
    if len(values) < 2:
        return values

    key_fn = key or (lambda value: value)
    keys = [key_fn(value) for value in values]
    left = 0
    right = len(values) - 1

    while left < right:
        min_index = left
        max_index = left

        for index in range(left + 1, right + 1):
            if _before(keys[index], keys[min_index], reverse):
                min_index = index
            if _before(keys[max_index], keys[index], reverse):
                max_index = index

        _swap(values, keys, left, min_index)

        if max_index == left:
            max_index = min_index

        _swap(values, keys, right, max_index)
        left += 1
        right -= 1

    return values


def _before(left, right, reverse):
    return left > right if reverse else left < right


def _swap(values, keys, left, right):
    if left != right:
        values[left], values[right] = values[right], values[left]
        keys[left], keys[right] = keys[right], keys[left]
